File: clients/swift/tests/SerializationTests.swift
import Foundation
import Testing
@testable import MyMindClient

@Suite("CreateObjectRequest serialization")
struct SerializationTests {
    private func encode(_ req: CreateObjectRequest) throws -> [String: Any] {
        let data = try JSONEncoder().encode(req)
        return try JSONSerialization.jsonObject(with: data) as! [String: Any]
    }

    @Test("Encodes url as string")
    func encodesUrl() throws {
        let req = CreateObjectRequest(url: URL(string: "https://example.com/page")!)
        let json = try encode(req)
        #expect(json["url"] as? String == "https://example.com/page")
    }

    @Test("Omits nil fields")
    func omitsNilFields() throws {
        let req = CreateObjectRequest(title: "Hello")
        let json = try encode(req)

        #expect(json["title"] as? String == "Hello")
        #expect(json["url"] == nil)
        #expect(json["content"] == nil)
        #expect(json["tags"] == nil)
        #expect(json["spaces"] == nil)
    }

    @Test("Excludes blob from JSON body")
    func excludesBlob() throws {
        let req = CreateObjectRequest(
            title: "Photo",
            blob: Blob(data: Data([0x89, 0x50, 0x4E, 0x47]), type: "image/png", name: "photo.png")
        )
        let json = try encode(req)

        #expect(json["blob"] == nil)
        #expect(json["title"] as? String == "Photo")
    }

    @Test("Encodes tags and spaces")
    func encodesTagsAndSpaces() throws {
        let req = CreateObjectRequest(
            tags: [ObjectTag(name: "reading")],
            spaces: [ObjectSpace(id: "sp_123")]
        )
        let json = try encode(req)

        let tags = json["tags"] as! [[String: Any]]
        let spaces = json["spaces"] as! [[String: Any]]
        #expect(tags[0]["name"] as? String == "reading")
        #expect(spaces[0]["id"] as? String == "sp_123")
    }
}

File: clients/csharp/tests/SerializationTests.cs
using System.Text.Json;
using MyMind.Actions;
using MyMind.Models;
using Xunit;

namespace MyMind.Tests;

public class SerializationTests
{
    [Fact]
    public void CreateObjectRequest_Serializes_Url_As_String()
    {
        var req = new CreateObjectRequest { Url = new Uri("https://example.com/page") };

        var json = JsonSerializer.Serialize(req);
        var doc = JsonDocument.Parse(json).RootElement;

        Assert.Equal("https://example.com/page", doc.GetProperty("url").GetString());
    }

    [Fact]
    public void CreateObjectRequest_Omits_Null_Fields()
    {
        var req = new CreateObjectRequest { Title = "Hello" };

        var json = JsonSerializer.Serialize(req);
        var doc = JsonDocument.Parse(json).RootElement;

        Assert.Equal("Hello", doc.GetProperty("title").GetString());
        Assert.False(doc.TryGetProperty("url", out _));
        Assert.False(doc.TryGetProperty("content", out _));
        Assert.False(doc.TryGetProperty("tags", out _));
        Assert.False(doc.TryGetProperty("spaces", out _));
    }

    [Fact]
    public void CreateObjectRequest_Excludes_Blob_From_Json()
    {
        var req = new CreateObjectRequest
        {
            Title = "Photo",
            Blob = new Blob
            {
                Stream = new MemoryStream([0x89, 0x50, 0x4E, 0x47]),
                Type = "image/png",
                Name = "photo.png",
            },
        };

        var json = JsonSerializer.Serialize(req);
        var doc = JsonDocument.Parse(json).RootElement;

        Assert.False(doc.TryGetProperty("blob", out _));
        Assert.Equal("Photo", doc.GetProperty("title").GetString());
    }

    [Fact]
    public void CreateObjectRequest_Serializes_Tags_And_Spaces()
    {
        var req = new CreateObjectRequest
        {
            Tags = [new ObjectTag { Name = "reading" }],
            Spaces = [new ObjectSpace { Id = "sp_123" }],
        };

        var json = JsonSerializer.Serialize(req);
        var doc = JsonDocument.Parse(json).RootElement;

        Assert.Equal("reading", doc.GetProperty("tags")[0].GetProperty("name").GetString());
        Assert.Equal("sp_123", doc.GetProperty("spaces")[0].GetProperty("id").GetString());
    }
}
